File: src/tools/shared/secretKeys.ts
import * as path from "node:path";
import type { StateStore } from "src/state/StateStore";
import { readEnvFile } from "src/state/envFile";
import { shouldRedact } from "src/state/secretRedactor";

export interface SecretKeysContext {
  cwd: string;
  stateStore: StateStore;
}

function resolveEnvFile(cwd: string, envFilePath: string): string {
  return path.isAbsolute(envFilePath) ? envFilePath : path.join(cwd, envFilePath);
}

/**
 * Collect known secret env keys for a stack so logs/output can be scrubbed.
 * Mirrors applyStack.ts:stackSecretKeys but sources the definition from state.
 * Sources:
 *   (a) `x-docker-agent.envFileSources[*].addedKeys` (unconditional),
 *   (b) per-service `environment` keys that look secret (shouldRedact),
 *   (c) per-service `env_file` keys (resolved against cwd) that look secret.
 */
export function collectSecretKeys(stackName: string, ctx: SecretKeysContext): Set<string> {
  const keys = new Set<string>();
  const def = ctx.stateStore.read(stackName);
  if (!def) return keys;

  for (const source of Object.values(def["x-docker-agent"].envFileSources)) {
    for (const key of source.addedKeys ?? []) keys.add(key);
  }
  for (const spec of Object.values(def.services)) {
    for (const key of Object.keys(spec.environment ?? {})) {
      if (shouldRedact(key)) keys.add(key);
    }
    for (const envFile of spec.env_file ?? []) {
      const values = readEnvFile(resolveEnvFile(ctx.cwd, envFile));
      for (const key of Object.keys(values)) {
        if (shouldRedact(key)) keys.add(key);
      }
    }
  }
  return keys;
}

File: servers/docker-mcp-server/src/docker_mcp_server/tools/shared/secret_keys.py
"""Collect known secret env keys for log scrubbing.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from docker_mcp_server.state.env_file import read_env_file
from docker_mcp_server.state.secret_redactor import should_redact
from docker_mcp_server.state.state_store import StateStore


@dataclass
class SecretKeysContext:
    cwd: str
    state_store: StateStore


def _resolve_env_file(cwd: str, env_file_path: str) -> str:
    path = Path(env_file_path)
    if path.is_absolute():
        return str(path)
    return str(Path(cwd) / env_file_path)


def collect_secret_keys(stack_name: str, ctx: SecretKeysContext) -> set[str]:
    """Collect secret env keys for a stack from state and env files."""
    keys: set[str] = set()
    definition = ctx.state_store.read(stack_name)
    if definition is None:
        return keys

    for source in definition.x_infra_agent.env_file_sources.values():
        for key in source.added_keys or []:
            keys.add(key)

    for spec in definition.services.values():
        for key in spec.environment or {}:
            if should_redact(key):
                keys.add(key)
        for env_file in spec.env_file or []:
            values = read_env_file(_resolve_env_file(ctx.cwd, env_file))
            for key in values:
                if should_redact(key):
                    keys.add(key)

    return keys


__all__ = ["SecretKeysContext", "collect_secret_keys"]
